django/curator/tag_deduplication.py

```diff
@@ -0,0 +1,161 @@
+import abc
+import os
+import re
+from typing import List
+
+import dedupe
+from taggit.models import Tag
+
+from curator.models import CanonicalTagMapping, CanonicalTag
+
+
+class TagPreprocessing:
+    def preprocess(tag_name: str) -> List[str]:
+        tag_names = re.split(r" and |\;|\,", tag_name)
+        tag_names = [tag_name.strip() for tag_name in tag_names]
+        return tag_names
+
+
+class AbstractTagDeduper(abc.ABC):
+    FIELDS = [{"field": "name", "type": "String"}]
+
+    # Gets the model's most uncertain pairs
+    def uncertain_pairs(self):
+        return self.deduper.uncertain_pairs()
+
+    # Pairs will be two tags.
+    # If the tags refer to the same thing, is_distinct = True
+    # Otherwise, is_distinct should be equal to False
+    def mark_pairs(self, pairs, is_distinct: bool):
+        labelled_examples = {"match": [], "distinct": []}
+
+        example_key = "distinct" if is_distinct else "match"
+        labelled_examples[example_key] = pairs
+
+        self.deduper.mark_pairs(labelled_examples)
+
+    def prepare_training_data(self):
+        tags = Tag.objects.all().values()
+        data = {row["id"]: row for i, row in enumerate(tags)}
+        return data
+
+    def console_label(self):
+        dedupe.console_label(self.deduper)
+
+
+# This class is used to help to make the initial canonical list.
+# Besides that, there isn't much other use for this.
+# If the curator can directly provide a small canonical list, even a small one will do, this class will not be needed.
+class TagClusterer(AbstractTagDeduper):
+    TRAINING_FILE = "curator/clustering_training.json"
+
+    def __init__(self, clustering_threshold):
+        self.clustering_threshold = clustering_threshold
+
+        self.deduper = dedupe.Dedupe(AbstractTagDeduper.FIELDS)
+        self.prepare_training()
+
+    # The training data is stored in a file.
+    # If it exists, load from the file
+    # Otherwise, start from scratch
+    def prepare_training(self):
+        data = self.prepare_training_data()
+        if os.path.exists(TagClusterer.TRAINING_FILE):
+            with open(TagClusterer.TRAINING_FILE, "r") as training_file:
+                self.deduper.prepare_training(data, training_file)
```

Comment:
Would it be better to store training data in the db instead of a file in the src tree? Looks like `Dedupe.prepare_training()` needs a file but this could be done by writing to an in-memory file object: https://docs.python.org/3/library/io.html#io.StringIO

Reply:
I am not too sure what the benefits/costs are for storing the training data in the db vs file system but maybe I can make a function for it in case it is needed?

Reply:
I'm OK with it living on the filesystem. It's essentially private / transient data that is only used once to bootstrap the gazetteering if I understand the process correctly...